utcp.c

```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+ *
+ * 0..3 represent the SACK entries. The ^ indicates up to which point we want
+ * to remove data from the receive buffer. The idea is to substract "len"
+ * from the offset of all the SACK entries, and then remove/cut down entries
+ * that are shifted to before the start of the receive buffer.
+ *
+ * There are three cases:
+ * - the SACK entry is ahead of ^, in that case just change the offset.
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
+ */
+static void sack_consume(struct utcp_connection *c, size_t len) {
+	debug("sack_consume %lu\n", (unsigned long)len);
+
+	buffer_get(&c->rcvbuf, NULL, len);
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
+		if(len < c->sacks[i].offset) {
+			c->sacks[i].offset -= len;
+			i++;
+		} else if(len < c->sacks[i].offset + c->sacks[i].len) {
+			c->sacks[i].len -= len - c->sacks[i].offset;
+			c->sacks[i].offset = 0;
+			i++;
+		} else {
+			if(i < NSACKS - 1) {
+				// move remaining sacks one ahead
+				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
+				// clean last sack len
+				c->sacks[NSACKS - 1].len = 0;
+			} else {
+				c->sacks[i].len = 0;
+				break;
+			}
+		}
+	}
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; i++)
+		debug("SACK[%d] offset %u len %u\n", i, c->sacks[i].offset, c->sacks[i].len);
+}
+
+static void handle_out_of_order(struct utcp_connection *c, uint32_t offset, const void *data, size_t len) {
+	debug("out of order packet, offset %u\n", offset);
+
+	// drop packets that are ahead of max buffer size
+	if(offset >= c->rcvbuf.maxsize) {
+		debug("warning: packet offset %u ahead of max buffer size %u\n", offset, c->rcvbuf.maxsize);
+		return;
+	}
+
+	// Packet loss or reordering occured. Store the data in the buffer.
+	ssize_t rxd = buffer_put_at(&c->rcvbuf, offset, data, len);
+
+	// Make note of where we put it.
+	for(int i = 0; i < NSACKS; i++) {
+		if(!c->sacks[i].len) { // nothing to merge, add new entry
+			debug("New SACK entry %d\n", i);
+			c->sacks[i].offset = offset;
+			c->sacks[i].len = rxd;
+			break;
+		} else if(offset < c->sacks[i].offset) {
+			if(offset + rxd < c->sacks[i].offset) { // insert before
+				if(!c->sacks[NSACKS - 1].len) { // only if room left
+					debug("Insert SACK entry at %d\n", i);
+					memmove(&c->sacks[i + 1], &c->sacks[i], (NSACKS - i - 1) * sizeof c->sacks[i]);
+					c->sacks[i].offset = offset;
+					c->sacks[i].len = rxd;
+				}
+				break;
+			} else { // merge
+				debug("Merge with start of SACK entry at %d\n", i);
+				c->sacks[i].offset = offset;
+				break;
+			}
+		} else if(offset <= c->sacks[i].offset + c->sacks[i].len) {
+			if(offset + rxd > c->sacks[i].offset + c->sacks[i].len) { // merge
+				debug("Merge with end of SACK entry at %d\n", i);
+				c->sacks[i].len = offset + rxd - c->sacks[i].offset;
+				// TODO: handle potential merge with next entry
```

Comment:
On Thu, Oct 29, 2015 at 06:27:56AM -0700, vollkorn1982 wrote:

> > +		} else if(offset <= c->sacks[i].offset + c->sacks[i].len) {
> > +			if(offset + rxd > c->sacks[i].offset + c->sacks[i].len) { // merge
> > +				debug("Merge with end of SACK entry at %d\n", i);
> > +				c->sacks[i].len = offset + rxd - c->sacks[i].offset;
> > +				// TODO: handle potential merge with next entry
> 
> How crucial is this? Maybe add some debug output to find out under which circumstances and how often this occurs?

Nothing will break by not handling this (I tested that), but it is more
efficient to do it properly. It might occur when there is a lot of
packet reordering. I'll do this TODO.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>